idea-plugin/src/com/reason/ide/annotations/ErrorAnnotator.java

```diff
@@ -1,113 +1,229 @@
+    Project project = psiFile.getProject();
+    VirtualFile sourceFile = psiFile.getVirtualFile();
 
-      int startOffset = editor.logicalPositionToOffset(start);
-      int endOffset = editor.logicalPositionToOffset(end);
-      if (0 < startOffset && 0 < endOffset && startOffset < endOffset) {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug("annotate " + startOffset + ":" + endOffset + " '" + message + "'");
-        }
-        TextRangeInterval range = new TextRangeInterval(startOffset - 1, endOffset - 1);
-        result.add(new BsbErrorAnnotation(info.isError, message, range, start));
-      } else {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug(
-              "Failed to locate info: "
-                  + start
-                  + "->"
-                  + end
-                  + ", offsets "
-                  + startOffset
-                  + "->"
-                  + endOffset
-                  + ", info "
-                  + info);
-        }
+    // create temporary compilation directory, once
+    if (m_compilationDirectory == null) {
+      m_compilationDirectory = createTempCompilationDirectory(project);
+    }
+
+    Optional<VirtualFile> contentRoot = BsPlatform.findContentRootForFile(project, sourceFile);
+    if (!contentRoot.isPresent()) {
+      LOG.info("Unable to find BuckleScript content root.");
+      return null;
+    }
+
+    VirtualFile bsExecutableDirectory = contentRoot.get().findFileByRelativePath("lib/bs");
+    if (bsExecutableDirectory == null) {
+      LOG.info("Unable to find lib/bs for BuckleScript project.");
+      return null;
+    }
+
+    // Read bsConfig to get the jsx value and ppx
+    // @TODO register a file listener for bsconfig.json
+    VirtualFile bsConfigFile = contentRoot.get().findFileByRelativePath("bsconfig.json");
+    if (bsConfigFile == null) {
+      LOG.info("No bsconfig.json found for content root: " + contentRoot);
+      return null;
+    }
+    BsConfig config = BsConfigReader.read(bsConfigFile);
+    String jsxVersion = config.getJsxVersion();
+    String namespace = config.getNamespace();
+
+    // If a directory is marked as dev-only, it won't be built and exposed to other "dev"
+    // directories in the same project
+    // https://bucklescript.github.io/docs/en/build-configuration#sources
+    // @TODO register a file listener for build.ninja
+    BsCompiler bucklescript = ServiceManager.getService(project, BsCompiler.class);
+    Ninja ninja = bucklescript.readNinjaBuild(contentRoot.get());
+    for (String devSource : config.getDevSources()) {
+      VirtualFile devFile = contentRoot.get().findFileByRelativePath(devSource);
+      if (devFile != null && FileUtil.isAncestor(devFile.getPath(), sourceFile.getPath(), true)) {
+        ninja.addInclude(devSource);
       }
     }
 
-    return result;
+    if (!BsPlatform.findBscExecutable(project, sourceFile).isPresent()) {
```

Comment:
I have the feeling we are doing a lot of this. maybe just keep the test in BscProcess and remove the others ?

Reply:
cool to do this in a separate PR? I'd like to limit the scope of changes here

Reply:
yes. but I'm just talking about that particular line, that could be removed.

Reply:
oh wow, yeah didn't realize we weren't using the executable's path. removed.